### PR TITLE
Remove EventLoop version of Endpoint Discovery code

### PR DESCRIPTION
**Pull request by adam-fowler**
This ended up being slightly more complex than the removal of the waiter and paginators EL code as I had to rewrite the endpoint discovery in Swift Concurrency.

 I have added a new type `ExpiringValue` which stores a value and an expiration date for that value. When you get the value you have to provide a function that would update the value if it is out of date. The type also ensures you aren't calling the update function while an update is already in progress (instead you wait on the result of the in progress update). This will be used in other parts of the codebase when I get to them eg some of the credential providers.

### CODE REVIEW

File: Sources/SotoCore/AWSEndpointDiscovery.swift [hunk 1]
```diff
@@ -39,96 +39,52 @@ public struct AWSEndpoints {
...
+public struct AWSEndpointStorage: Sendable {
+    let endpoint: ExpiringValue<String>
 
-    /// Initialize endpoint storage
-    /// - Parameter endpoint: Initial endpoint to use
-    public init(endpoint: String) {
-        self.expiration = Date.distantPast
-        self.endpoint = endpoint
+    public init() {
+        self.endpoint = .init(threshold: 3 * 60)
```
Comment (0xTim) [inline Sources/SotoCore/AWSEndpointDiscovery.swift:47]: Is there a reason why this is hard coded as 180?

File: Sources/SotoCore/Concurrency/ExpiringValue.swift [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import Foundation
+import Logging
+
+/// Type holding a value and an expiration value.
+///
+/// When accessing the value you have to provide a closure that will update the
+/// value if it has expired or is about to expire. The type ensures there is only
+/// ever one value update running at any one time. If an update is already running
+/// when you call `getValue` it will wait on the current update function to finish.
+actor ExpiringValue<T> {
```
Comment (0xTim) [inline Sources/SotoCore/Concurrency/ExpiringValue.swift:24]: Does this need to be public?

Comment (adam-fowler) [reply Sources/SotoCore/Concurrency/ExpiringValue.swift:24]: Not at this point. There is a possibility I will use it in some Cognito credential providers. But for the moment I am only using it internally

File: Sources/SotoCore/AWSEndpointDiscovery.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/SotoCore/AWSEndpointDiscovery.swift:48]: It is set to 3 minutes. Basically when an endpoint is about to expire it has 3 minutes to get a new one. I didn't think this was something that needed a configurable value. I've added a comment

**Review by 0xTim — APPROVED** [review]

**Pull request merged**